### DISCUSSION TITLE
Semantic Highlighting in Injected Language

### DISCUSSION THREAD

**Discussion by SteveBenz** [microsoft/vscode-discussions#130] [Extension Development QnA] [answered]
I'm trying to create an extension that (as a first step) will look at a comment and figure out if it's worth reading or not.  (E.g. perhaps it contains almost exclusively words that are found in the following line or two of code).  First off, am I even approaching this in the best way?

From reading the docs, I think the way to do it is to create a "language" with a semantic token analyzer, and then have that be an "injected language" that is scoped to comments in real languages.  That'd report the block back either as a plain comment or a "useless" comment, which would render less visibly than a plain comment.

If I'm on the right track with that, then I'm stuck on the part of getting the injected language to actually run the semantic token analyzer.

```json
  "contributes": {
    "languages": [
      {
        "id": "commentContents",
        "extensions": [ ".comcon" ]
      }
    ],
    "grammars": [
      {
        "path": "./syntaxes/injection.json",
        "scopeName": "comment-rater.injection",
        "injectTo": ["source.js"],
        "embeddedLanguages": {
          "meta.embedded.inline.stuff": "commentContents"
        }
      }
    ]
  },
```
And this injection.json:

```json
{
    "scopeName": "comment-rater.injection",
    "injectionSelector": "L:comment",
    "patterns": [
        {
            "include": "#everything"
        }
    ],
    "repository": {
        "everything": {
            "match": ".*",
            "name": "meta.embedded.inline.stuff"
        }
    }
}
```
extensions.ts contains this:

```ts
export function activate(context: vscode.ExtensionContext) { ...
	context.subscriptions.push(
		vscode.languages.registerDocumentSemanticTokensProvider(
			{ language: 'commentContents' },
			new DocumentSemanticTokensProvider(),
			legend));
... }

class DocumentSemanticTokensProvider implements vscode.DocumentSemanticTokensProvider {
	async provideDocumentSemanticTokens(document: vscode.TextDocument, token: vscode.CancellationToken): Promise<vscode.SemanticTokens> { ... }
}
```

Each piece works separately -- that is, I know the injection is working because comments in JavaScript files turn white (because the syntax parser converted them to a `meta.embedded.inline.scope` scope.)  I know my semantic token provider works because when I open a file that ends with ".comcon" I see highlighting happening.

What doesn't happen is that my DocumentSemanticTokensProvider isn't getting called when I look at JavaScript files with comments in them.

Any clues greatly appreciated!

**Comment by pgfearo** [comment, marked as answer]
I think you're anticipating functionality that isn't quite there yet. See [#81558](https://github.com/microsoft/vscode/issues/81558) in the VSCode repo.

**Reply by SteveBenz** [reply]
Before I give up on this avenue, I want to make sure that you mean you're dead sure it can't be done...  When I read this:

https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide

Particularly the section entitled "EMBEDDED LANGUAGES", I just can't find a way to interpret the purpose of `contributes.grammars.embeddedlanguages` any other way.  (If that section had included the injection.json that went along with it, that would have made it a lot clearer.)  I think if I could understand that section better, I'd know why things weren't working for me.  It also seems surprising that this isn't a working feature of VSCode given that systems that mix HTML and procedural languages are as common as dirt.  (Maybe all of them combine an HTML grammar with Javascript/whatever into one language server?)

However, the issue you cite combined with the fact that all of the documentation on embedded languages falls under the "Syntax Highlight Guide" page and none of it under the "Semantic Highlight Guide" makes me think that you're correct that the capability isn't there.

I just don't want to give up on the idea until I'm sure it's not going to be feasible.

**Reply by aeschli** [reply]
@pgfearo is right. The semantic highlight provider always works on a full document and is only invoked when its selector matches the top-level language mode of the document.
That's the case with all providers (code completion, hovers, ....).
The 'embeddedLanguages` scopes from the `grammar` contribution point are used with basic language features such as bracket matching, autoClosingPairs, comment tokens....

**Reply by SteveBenz** [reply]
Thanks for providing certainty here.  I'll add a +1 onto #81558 unless there's some other place where this feature request is being tracked...

**Reply by SteveBenz** [reply]
Thanks.  I'll look into other avenues.  Perhaps something along the lines of a linter...

**Comment by isidorn** [comment]
fyi @aeschli for semantic highlighting
